consensus/istanbul/core/commit.go

```diff
@@ -207,62 +218,135 @@ func (c *core) handleCheckedCommitForCurrentSequence(msg *istanbul.Message, comm
+		if istanbul.IsLastBlockOfEpoch(proposal.Number().Uint64(), c.config.Epoch) {
+			epochBitmap, epochAggregate, err := AggregateSeals(
+				c.current.Commits(),
+				func(c *istanbul.CommittedSubject) []byte { return c.EpochValidatorSetSeal },
+			)
+			if err != nil {
+				nextRound := new(big.Int).Add(c.current.Round(), common.Big1)
+				logger.Warn("Error on commit, waiting for desired round", "reason", "getAggregatedSeal", "err", err, "desired_round", nextRound)
+				c.waitForDesiredRound(nextRound)
+				return nil
+			}
+
+			aggregatedEpochValidatorSetSeal.Bitmap = epochBitmap
+			aggregatedEpochValidatorSetSeal.Signature = epochAggregate
 		}
 
-	} else if c.current.GetPrepareOrCommitSize() >= minQuorumSize && c.current.State().Cmp(StatePrepared) < 0 {
-		err := c.current.TransitionToPrepared(minQuorumSize)
+		logger.Trace("Got a quorum of commits", "tag", "stateTransition", "commits", commits)
+		err = c.commit(aggregatedSeal, aggregatedEpochValidatorSetSeal)
 		if err != nil {
-			logger.Error("Failed to create and set prepared certificate", "err", err)
+			logger.Error("Failed to commit()", "err", err)
 			return err
 		}
-		// Process Backlog Messages
-		c.backlog.updateState(c.current.View(), c.current.State())
-
-		logger.Trace("Got quorum prepares or commits", "tag", "stateTransition", "commits", c.current.Commits, "prepares", c.current.Prepares)
-		c.sendCommit()
 	}
 	return nil
 
 }
 
+// generateAggregateCommittedSeal will generate the aggregate committed seal
+// verify it and return it.
+func (c *core) generateAggregateCommittedSeal() (types.IstanbulAggregatedSeal, error) {
+	commits := c.current.Commits()
+	if commits.Size() < c.current.ValidatorSet().MinQuorumSize() {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("insufficient commits to construct aggregate seal")
+	}
+	bitmap, aggregate, err := AggregateSeals(
+		commits,
+		func(c *istanbul.CommittedSubject) []byte { return c.CommittedSeal },
+	)
+	if err != nil {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("failed to aggregate seals: %v", err)
+	}
+
+	aggregatedSeal := types.IstanbulAggregatedSeal{Bitmap: bitmap, Signature: aggregate, Round: c.current.Round()}
+	err = c.backend.VerifyAggregatedSeal(c.current.Proposal().Hash(), c.current.ValidatorSet(), aggregatedSeal)
+	if err != nil {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("failed verify aggregate seal: %v", err)
+	}
+	return aggregatedSeal, nil
+}
+
+// removeInvalidCommittedSeals individually verifies the committed seal on each
+// commit message and discards commits with invalid committed seals.
+//
+// Note that when commit messages are discarded they are not removed from the
+// round state database, this should not pose a problem however because if this
+// step is reached again they will again be discarded.
+func (c *core) removeInvalidCommittedSeals() {
+
+	commits := c.current.Commits()
+
+	for _, msg := range commits.Values() {
```

Comment:
Maybe worth doing a binary search / parallelizing this.

Reply:
In rank of the optimizations I'd like to see implemented, we should consider:
1. Streaming aggregation
2. Binary search for bad signatures
3. Parallization

I'd say 1 is most important to implement now, as it implies non-trivial code changes to how this feature works. 2 and 3 should be fairly easy though and are self-contained, so it would be nice to implement those as well.

Reply:
@nategraf Just to clarify, streaming aggregation is not a performance optimisation beyond what has been done here, it would probably cause a minor reduction in performance but would let us know that all signatures are individually valid?

Reply:
@nategraf I'm going to see about changing this to the streaming model.

Reply:
Retraction on the statement of parallization being easy. As we discussed elsewhere, we probably shouldn't open that can of worms right now.